src/ability/msg.rs

```diff
@@ -0,0 +1,183 @@
+use crate::{ability::traits::Command, prove::TryProve};
+use libipld_core::ipld::Ipld;
+use std::{collections::BTreeMap, str::FromStr};
+use url::Url;
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct Msg {
+    to: Url,
+    from: Url,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct MsgSend {
+    to: Url,
+    from: Url,
+    message: String,
+}
+
+// TODO is the to or from often also the subject? Shoudl that be accounted for?
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct MsgReceive {
+    to: Url,
+    from: Url,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct MsgReceiveBuilder {
+    to: Option<Url>,
+    from: Option<Url>,
+}
+
+impl From<MsgReceive> for MsgReceiveBuilder {
+    fn from(msg: MsgReceive) -> Self {
+        Self {
+            to: Some(msg.to),
+            from: Some(msg.from),
+        }
+    }
+}
+
+impl TryFrom<MsgReceiveBuilder> for MsgReceive {
+    type Error = MsgReceiveBuilder;
+
+    fn try_from(builder: MsgReceiveBuilder) -> Result<Self, MsgReceiveBuilder> {
+        // FIXME
+        if let (Some(to), Some(from)) = (builder.clone().to, builder.clone().from) {
+            Ok(Self { to, from })
+        } else {
+            Err(builder.clone()) // FIXME
+        }
+    }
+}
+
+impl From<MsgReceive> for Ipld {
+    fn from(msg: MsgReceive) -> Self {
+        let mut map = BTreeMap::new();
```

Comment:
Nitpick: for cases like this where the set of keys is statically known, it can sometimes be a little less verbose to use the `FromIterator` trait to construct maps, and avoid needing to mutate the map key-by-key:

```rust
BTreeMap::from_iter([
  ("to".to_string(), msg.to.to_string().into()),
  ("from".to_string(), msg.from.to_string().into())
]).into()
```

Reply:
Great tip! Thanks :)